opm/autodiff/WellStateFullyImplicitBlackoil.hpp

```diff
@@ -207,47 +221,17 @@ namespace Opm
              * representations.
              */
 
-            for( auto w = 0; w < nw; ++w ) {
-                using rt = data::Rates::opt;
-                std::map< size_t, data::Completion > completions;
-
-                // completions aren't supported yet
-                //const auto* begin = wells_->well_connpos + w;
-                //const auto* end = wells_->well_connpos + w + 1;
-                //for( auto* i = begin; i != end; ++i ) {
-                //    const auto perfrate = this->perfPhaseRates().begin() + *i;
-                //    data::Rates perfrates;
-                //    perfrates.set( rt::wat, *(perfrate + 0) );
-                //    perfrates.set( rt::oil, *(perfrate + 1) );
-                //    perfrates.set( rt::gas, *(perfrate + 2) );
-
-                //    const size_t active_index = wells_->well_cells[ *i ];
+            for( const auto& wt : this->wellMap() ) {
+                const auto w = wt.second[ 0 ];
+                auto& well = res.at( wt.first );
+                well.control = this->currentControls()[ w ];
 
-                //    completions.emplace( active_index,
-                //        data::Completion{ active_index, perfrates } );
-                //}
-
-                const auto wellrate_index = np * w;
-                const auto& wv = this->wellRates();
-
-                data::Rates wellrates;
-                if( pu.phase_used[BlackoilPhases::Aqua] ) {
-                    wellrates.set( rt::wat, wv[ wellrate_index + pu.phase_pos[BlackoilPhases::Aqua] ] );
-                }
-
-                if( pu.phase_used[BlackoilPhases::Liquid] ) {
-                    wellrates.set( rt::oil, wv[ wellrate_index + pu.phase_pos[BlackoilPhases::Liquid] ] );
-                }
-
-                if( pu.phase_used[BlackoilPhases::Vapour] ) {
-                    wellrates.set( rt::gas, wv[ wellrate_index + pu.phase_pos[BlackoilPhases::Vapour] ] );
+                for( auto& cpair : well.completions ) {
+                    const auto rates = this->perfPhaseRates().begin() + (np * w);
```

Comment:
Incorrect index. Here you must use `np * wt.second[1]` not `np * wt.second[0]` (as this is equivalent to) to get an iterator to the *start of* a well's perforation data. Then you must add `np * local_comp_index` to get the data for the correct index, where `local_comp_index` is the index of the completion within the set of completions for that single well (i.e. 0...(nwc-1), where nwc == `wt.second[2]` == the number of completions for the well in question).